### ISSUE TITLE
An endless call when run zone.run in ApplicationRef.isStable

### ISSUE THREAD

**Issue by dadalxx** [angular/angular#49275]
### Which @angular/* package(s) are the source of the bug?

core

### Is this a regression?

Yes

### Description

I followed the the guide about [ApplicationRef#is-stable-examples](https://angular.io/api/core/ApplicationRef#is-stable-examples). When i tried to delete the pipe like this, and the console.log executed endless until the browser crushed.

```
constructor(appRef: ApplicationRef), zone: NgZone)) {
  appRef.isStable.subscribe(state => zone.run(() => {
    console.log('endless call');
    // You don't want to do this
    ngZone.run(() => (this.value = state));
  }));
}
```

It is horrible to misuse it. Maybe `ApplicationRef`, `ngZone` or `zone.js` could take some protection to avoid this.  
Or at least, adding a warning in the doc.

### Please provide a link to a minimal reproduction of the bug

https://stackblitz.com/edit/angular-hcxfu5?file=src%2Fmain.ts

### Please provide the exception or error you saw

```true
The browser crushed when run `ngZone.run` in the subscription of `ApplicationRef.isStable`.
```

### Please provide the environment you discovered this bug in (run `ng version`)

```true
Angular CLI: 14.2.10
Node: 16.15.0
Package Manager: npm 8.5.5
OS: win32 x64

Angular: 14.2.0+sha-e811626-with-local-changes
... core

Package                             Version
-------------------------------------------------------------
@angular-devkit/architect           0.1402.10
@angular-devkit/build-angular       14.2.10
@angular-devkit/core                14.2.10
@angular-devkit/schematics          14.2.10
@angular/animations                 14.2.0
@angular/cli                        14.2.10
@angular/common                     14.2.0
@angular/compiler                   14.2.0
@angular/compiler-cli               14.2.0
@angular/forms                      14.2.0
@angular/platform-browser           14.2.0
@angular/platform-browser-dynamic   14.2.0
@angular/router                     14.2.0
@schematics/angular                 14.2.10
rxjs                                7.5.7
typescript                          4.7.4
```

### Anything else?

Otherwise, still in the doc of  [is-stable-examples](https://angular.io/api/core/ApplicationRef#is-stable-examples). I saw `cd.detectChanges()` is used in constructor from the example. Whether it is a bad example, cause it would cause  a `ASSERTION ERROR`.

**Comment by alxhub** [comment]
The example you mention [works for me](https://stackblitz.com/edit/angular-ztrmsu?file=src%2Fmain.ts).

Calling `NgZone.run` unconditionally from `isStable` is definitely an infinite loop - `NgZone.run` adds a task to the zone, making it unstable and then stable again, which will run the `subscribe` callback again, which runs `NgZone.run` again, etc. This behavior is expected, and the examples don't suggest to me that it would behave otherwise.

**Comment by dadalxx** [comment]
I agree with you, and I read [checkStable](https://github.com/angular/angular/blob/e1355e7d16337177866cc1f8654b76381f26dae3/packages/core/src/zone/ng_zone.ts#L322) in ngZone which has the same hidden peril like this.

Maybe Angular could stop applying the stable observable or export a piped observable with `first` operation with rxjs.

**Comment by dadalxx** [comment]
Hi @JiaLiPassion 

I’ve saw your MR with moving the `zone._nesting--;`. But in your fix, the endless-call problem is only eliminated in `NgZone`. I have tried to run ngzone.run in the subscription of `NgZone.onStable`, the endless-call gone.

But, when i run ngzone.run in the subscription of `ApplicationRef.isStable`, the endless-call is still there. So I think it should fixed by both NgZone and ApplicationRef.

And Hi @alxhub @atscott 

I hope u can reevaluate this issue please.

- Make 2 PR to follow NgZone and ApplicationRef situation
- It's not only a doc problem, i think `label: core` could be added at least.
- Raise P5 to P4 or above

Thx!

**Comment by JeanMeche** [comment]
Closing as this is something that we likely won't address.
